File: apps/seller-web/src/lib/commune-rules.test.ts
import { describe, expect, it } from 'vitest';
import { communeHint, communeRequired, retainedCommuneId } from './commune-rules';

describe('communeRequired — mirrors the API rule', () => {
  it('is required only once a non-empty library has loaded', () => {
    expect(communeRequired({ loaded: true, communeCount: 6 })).toBe(true);
    // A city without communes (e.g. Likasi today) accepts the city alone.
    expect(communeRequired({ loaded: true, communeCount: 0 })).toBe(false);
    // Never claim optional while loading — the API decides.
    expect(communeRequired({ loaded: false, communeCount: 0 })).toBe(false);
  });
});

describe('retainedCommuneId — a city change clears an invalid commune', () => {
  const lubumbashi = ['c-kampemba', 'c-kenya', 'c-katuba'];
  it('keeps a commune of the new list, clears a foreign or empty one', () => {
    expect(retainedCommuneId('c-kenya', lubumbashi)).toBe('c-kenya');
    expect(retainedCommuneId('c-dilala', lubumbashi)).toBe('');
    expect(retainedCommuneId('', lubumbashi)).toBe('');
    expect(retainedCommuneId('c-kenya', [])).toBe('');
  });
});

describe('communeHint', () => {
  it('walks the states in order', () => {
    expect(communeHint(false, { loaded: false, loading: false, communeCount: 0 })).toBe('Sélectionnez d’abord une ville');
    expect(communeHint(true, { loaded: false, loading: true, communeCount: 0 })).toBe('Chargement des communes…');
    expect(communeHint(true, { loaded: true, loading: false, communeCount: 0 })).toBe(
      'Aucune commune enregistrée pour cette ville pour le moment',
    );
    expect(communeHint(true, { loaded: true, loading: false, communeCount: 3 })).toBe('Sélectionnez votre commune');
  });
});

File: apps/seller-mobile/test/core/utils/commune_rules_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:seller_mobile/core/utils/commune_rules.dart';

void main() {
  group('communeRequired — mirrors the API rule', () {
    test('required only once a non-empty library has loaded', () {
      expect(communeRequired(loaded: true, communeCount: 6), isTrue);
      expect(communeRequired(loaded: true, communeCount: 0), isFalse,
          reason:
              'a city without communes (e.g. Likasi today) accepts the city alone');
      expect(communeRequired(loaded: false, communeCount: 0), isFalse,
          reason: 'never claim optional while loading — the API decides');
    });
  });

  group('retainedCommuneId — city change clears an invalid commune', () {
    const lubumbashi = ['c-kampemba', 'c-kenya', 'c-katuba'];
    test('keeps a commune that belongs to the new list', () {
      expect(retainedCommuneId('c-kenya', lubumbashi), 'c-kenya');
    });
    test('clears a commune of another city, an empty or null selection', () {
      expect(retainedCommuneId('c-dilala', lubumbashi), isNull);
      expect(retainedCommuneId('', lubumbashi), isNull);
      expect(retainedCommuneId(null, lubumbashi), isNull);
      expect(retainedCommuneId('c-kenya', const <String>[]), isNull);
    });
  });

  group('communeHint', () {
    test('walks the states in order', () {
      expect(
          communeHint(
              cityChosen: false,
              loading: false,
              loaded: false,
              communeCount: 0),
          'Sélectionnez d’abord une ville');
      expect(
          communeHint(
              cityChosen: true, loading: true, loaded: false, communeCount: 0),
          'Chargement des communes…');
      expect(
          communeHint(
              cityChosen: true, loading: false, loaded: true, communeCount: 0),
          'Aucune commune enregistrée pour cette ville pour le moment');
      expect(
          communeHint(
              cityChosen: true, loading: false, loaded: true, communeCount: 3),
          'Sélectionnez votre commune');
    });
  });
}
